### DISCUSSION TITLE
I want to learn programming. Where should I start?

### DISCUSSION THREAD

**Discussion by chjch** [SERVIR-WA/GALUP#63] [Ideas]
Hi all, 

I want to start a conversation on this topic (as suggested in the title). This conversation will not be about programming in general but with **a focus on GIS-related programming and scripting**.  With that being said, I think the **first question** should be which language should I start with. Given that Python is the scripting language for both ArcGIS and QGIS and it is relatively easy for starters, I would say **_Python_** would be a great choice. 

I will put more information below and you are welcome to participate in this conversation.

Cheers,
Changjie

**Comment by Achidago** [comment]
This will be great. Looking forward to it

**Comment by Sheila6868** [comment]
Interesting. I look forward to it

**Comment by chjch** [comment]
To learn programming in GIS with _Python_, you need to learn about the **basics of Python** first. For that, I recommend you this [YouTube channel](https://www.youtube.com/channel/UCCezIgC97PvUuR4_gbFUs5g) by Corey Schafer. I think he is a great instructor and he's good at break things down. I personally learned many things from him.

First of all, I recommend you check out his series on [Python Programming Beginner Tutorials](https://www.youtube.com/playlist?list=PL-osiE80TeTskrapNbzXhwoFUiLCjGgY7). I encourage you to finish the whole series. But, the first 10 videos of this playlist would be an essential training for you to get a basic understanding of Python including **data types**, **conditional statements**, **looping**, **functions**, etc.

Secondly, you need the right tool to do the right things. When it comes to programming, the most important tool is the application software you code with, i.e., what we call the **_Integrated Development Environment_** (IDE). The IDE that I recommend you to use is Visual Studio Code, or VS Code. I consider it is the best text editor out there as of today. I recommend you checkout another video by Corey Schafer on [Setting up a Python Development Environment and Complete Overview](https://www.youtube.com/watch?v=-nh9rCzPJ20). The _Beginner Tutorials_ mentioned above was done in a different text editor called Sublime, which is not completely free. Contrarily, VS Code is completely free and gives you more power and still maintains lightweight (not as light as Sublime though). 

:warning: **REQUIREMENT**
1. Python: You already have Python 3.7 installed during the QGIS installation. The first video of the _Beginner Tutorials_ goes through the installation of Python for both Mac OS and Windows. But, we can just simply use what is already being installed.
2. VS Code: Before you follow the _Beginner Tutorials_, definitely check out the other video on setting up **VS Code**. Once you have VS Code correctly configured, then you can use it and code along with the _Beginner Tutorials_.

That will be all for now. I'll talk more about programming related to GIS next time.

:smiley: Cheers,
Changjie

**Reply by TrueKing** [reply]
Thank you for sharing. I am really interested in this thread since I am quite the beginner in this environment.

**Comment by Achidago** [comment]
Thanks for sharing!

**Comment by Sheila6868** [comment]
Thanks.

**Comment by ElvisMensah** [comment]
Great stuff

**Comment by Nyansafo** [comment]
thanks

**Comment by abigail93** [comment]
Thanks
